openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfRendererBuilder.java

```diff
@@ -41,6 +41,15 @@ public void run() throws Exception {
 		}
 	}
 
+	/**
+	 * Returns an instance of PdfBoxUserAgentFactory in case you<br/>
+	 * need to use custom implementation of {@link PdfBoxUserAgent} class
+	 * @return PdfBoxUserAgentFactory instance
+	 */
+	public PdfBoxUserAgentFactory getPdfBoxUserAgentFactory() {
+		return PdfBoxUserAgentFactory.instance();
```

Comment:
Thats bad API design. It also breaks the builder contract (i.e. everything but build() *always* returns PdfRendererBuilder for method chaining). Better would be 

- Just define the UserAgentFactory interface (as above)
- The builder should have a field userAgentFactory with a default implementation assign i.e.
```
        _userAgentFactory = new UserAgentFactory() {
            public PdfBoxUserAgent getPdfBoxUserAgent(PdfBoxOutputDevice outputDevice) {
                return new PdfBoxUserAgent(outputDevice);
            }
        };
```
- A setter to override the default implementation. i.e.
```
public PdfRendererBuilder setUserAgentFactory(UserAgentFactory customFactory) {
  this._userAgentFactory = customFactory;
  return this;
}
```
Of course it would require an additional parameter to the PdfBoxRenderer....

Reply:
I have a suspicion that you just did not understand the code above. 
"build() always returns PdfRendererBuilder for method chaining"
Of course it returns always a PdfRendererBuilder, it should be used in this manner:

        PdfBoxRenderer renderer = null;
        try {
            builder.getPdfBoxUserAgentFactory().registerGlobalPdfBoxUserAgentFactory(new PdfBoxUserAgentFactory.UserAgentFactory() {
                @Override
                public PdfBoxUserAgent getPdfBoxUserAgent(PdfBoxOutputDevice pdfBoxOutputDevice) {
                    return new ExtendedPdfBoxUserAgent(pdfBoxOutputDevice);
                }
            });
            renderer = builder.buildPdfRenderer();
            //customUserAgent(renderer);
            renderer.layout();
            renderer.createPDF();
        } finally {
            if (renderer != null)
                renderer.close();
        }

Reply:
@delafer  Sorry, but where is the ```unregisterPdfBoxUserAgentFactory()``` call? Oh, you forgot it. So a class loader leak is incoming on WAR reload ... This is what I mean by bad API design, if it's easy to forget something you *will* forget to do it... and the problem will only manifest after the n-th reload of the WAR..

Yes, creating an idiot proof API is hard, but should be the goal never the less.

Method chaining means: ```builder.setXY().setAbc().setWhatever().build()```. For another example look here: https://google.github.io/gson/apidocs/com/google/gson/GsonBuilder.html